docs/proposal-platforms.md

```diff
@@ -0,0 +1,312 @@
+# Proposal - initial support for platforms
+
+This will close [cylc-flow #3421](https://github.com/cylc/cylc-flow/issues/3421)
+and partially address
+[cylc-flow #2199](https://github.com/cylc/cylc-flow/issues/2199).
+
+## Key functionality
+
+* Introduce configuration and logic to recognise compute platforms in place of
+  task hosts. This will resolve a number of issues.
+  * In particular, this allows a compute cluster which can be accessed via
+    multiple login nodes (hosts) to be treated as logical unit.
+  * We will be able to configure multiple platforms which share the same hosts
+    (including `localhost`).
+
+* If a platform has a list of hosts then the host used for each operation on that
+  platform (job submission, polling, log retrieval, etc) will be randomised. This
+  should help balance the load. If a host becomes unavailable, try another random
+  host from the list, until exhausted. This will handle hosts becoming
+  unavailable after a job is submitted which is a major limitation of the current
+  logic.
+
+* It will be possible to specify the platform for a task as a platform alias
+  which defines a list of platforms. For example, you may have 2 different HPC
+  systems (configured as separate "platforms") and you are happy for a task to
+  run on either. The list will be randomised and the platforms tried in turn
+  until successful job submission.
+  * Note that the platform will be selected each time a job is run. If you
+    specify a platform alias which defines a list of platforms, tasks must not
+    assume that retries of that task will use the same platform.
+
+* It will be possible to specify the platform as a command in the same way as we
+  can currently specify hosts as a command. The command must return a valid
+  platform (or platform alias). We hope that this functionality will seldom be
+  needed in the future (the main use at present is to select from a list of login
+  nodes). However, there are still cases where this may be needed. For example
+  running a command which:
+  * Returns the platform which has the shortest queue time.
+  * Chooses the platform according to the day of the week (perhaps you have a
+    platform which can only be used at weekends?).
+
+* Note: it will **not** be possible to specify the platform as an environment
+  variable in the same way as we can currently specify remote hosts as an
+  environment variable (e.g. `platform = $ROSE_ORIG_HOST`).
+  * Unclear why we need this given the Jinja2 variables can be used?
+  * Using `$ROSE_ORIG_HOST` relies on Rose specifying extra options to set this
+    in `[cylc][environment]` which we should avoid if possible.
+
+## Example platform configurations
+
+```ini
+[platforms]
+    [[desktop\d\d|laptop\d\d]]
```

Comment:
I don't like the use of regexes for section headings due to the difficulties of overriding, however, this functionality will probably only really be used at a site level.

As a bare minimum this would be more flexible:

```ini
[[desktop\d\d, laptop\d\d]]
```

Reply:
I think sticking with a single regex is clearer. Hopefully it won't need to be heavily used.

Reply:
I tend to agree with Oliver here, in that we use comma-separated heading lists elsewhere (e.g. under [runtime] in suite.rc)

Reply:
Given that the configs can be templated with Jinja (even the global one), is there a benefit of having regex platform names? Seems like this brings additional rules and logic (which the user must be aware of in some contexts) compared to a simple pick a name from a list or group names under an alias.

In terms of extensions, instead of regex substitution in host names, a more general solution could be to perform runtime (or rather submission-time) templating of settings (https://github.com/cylc/cylc-flow/issues/2415#issuecomment-428552084). This would allow to not only alter the host name in arbitrary ways, but also other settings like job directives, which many users would welcome. I'm not huge fan of Jinja templating, but in this context I think it would work quite well (templating values of settings has much more limited scope than 'freestyle' templating an entire config file).

Reply:
(@wxtim can you note down @TomekTrzeciak 's suggestion here for consideration, particular the extension to job directives ...)

Reply:
The directives bit is already noted in cylc/cylc-flow#2415.

Regarding the regex, it's a trade off. Using a regex allows us to define a large number of platforms in a single section. We could define each platform separately using Jinja but we would end up with a potentially very large config which I don't like the idea of (for instance, I don't want `cylc get-global config` to return a massive long config).

Reply:
Fair enough, let's stick with regex then, for this.

Reply:
But, I still think it should be a comma-separated list - to keep each regex simpler, and for consistency with our support for comma-separated lists in other similar contexts.

Reply:
Another thought on this. With templating of settings one could cut down on the number of definitions with something like this:

```
#!jinja2
[platforms]
    [[boxNo]]
        host = {% raw %}{{ platform + "%02d" % range(1, 51)|random }}{% endraw %}
```

This would choose at random between `boxNo01` to `boxNo50` every time the host template is evaluated. It goes without saying that with a site installed jinja2 filters arbitrary logic can be performed this way.

Reply:
Good point (Sorry for serially agreeing with everyone - just no time to consider this in detail right now!).  @dpmatthews  @wxtim @oliver-sanders - we do need to consider templating more widely, and if we have it for other reasons then maybe no need for platform name regexes as well.  Disclaimer: I have not sat down to think through the pros and cons yet.